### DISCUSSION TITLE
Building firmware using ESP-IDF from inside a repo with absolute path containing spaces

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#10285] [ESP32] [answered]
Hello,

is there a way to build micropython for an ESP32 board from a directory which absolute path contains spaces?

I can build it from a directory not containing spaces, but I would like to use the esp-idf repo as a submodule to all my related projects (like I do with the MicroPython repo as submodule)... 

I tried all kinds of things for escaping the spaces but no luck so far.

Here is my setup script which I run from the repository root:
```
#!/bin/sh -e
REPO_ROOT="/Users/sarav/Sara Verbeecke Dropbox/My GitLab/micro_robot"

git submodule add https://github.com/micropython/micropython .submodules/micropython
git submodule add https://github.com/espressif/esp-idf .submodules/esp-idf

# Build the MicroPython UNIX port
cd "$REPO_ROOT/.submodules/micropython"
make -C mpy-cross
cd ports/unix
make submodules
make

# Build ESP-IDF
cd "$REPO_ROOT/.submodules/esp-idf"
git checkout v4.4
git submodule update --init --recursive
./install.sh
```

To build the firmware I use:
```
#!/bin/sh -e
REPO_ROOT="/Users/sarav/Sara Verbeecke Dropbox/My GitLab/micro_robot"
cd "$REPO_ROOT/.submodules/esp-idf"
. ./export.sh

cd "$REPO_ROOT/.submodules/micropython/ports/esp32"
make submodules
make BOARD=GENERIC_S3_SPIRAM
```
The last **make** command fails because it contains spaces...
I tried escaping them and manually pass the IDF_PATH (with escaped spaces) prior running the command but no luck. The ESP-IDF docs at [https://docs.espressif.com/projects/esp-idf/en/v4.4.3/esp32s3/api-guides/build-system.html](url) state 

> COMPONENT_DIR: The component directory. Evaluates to the absolute path of the directory containing CMakeLists.txt. The component path cannot contain spaces. This is the same as the CMAKE_CURRENT_SOURCE_DIR variable.

Any tips or is it a rabbit hole...🐇
(I'm working on a unix machine)

**Comment by karfas** [comment, marked as answer]
I'm pretty sure trying to fix this is waste of time. 
The EspressIF people have had months or years for the transition to CMake and obviously came up with the documentation part you cited instead of a fix.
Besides that, this looks more like an ESP-IDF issue, not one on the micropython side.

**Reply by saraverbeecke** [reply]
Putting everything located in absolute path containing no spaces is currently the only solution that worked for me. @sosi-deadeye, @Josverl thanks for your inputs, but no luck here (working on macOS)

**Comment by sosi-deadeye** [comment]
You can use a workaround with mount bind:
```
mount -o bind "/Users/sarav/Sara Verbeecke Dropbox/My GitLab/micro_robot" "/Users/sarav/micro_robot"
# or
mount -B "/Users/sarav/Sara Verbeecke Dropbox/My GitLab/micro_robot" "/Users/sarav/micro_robot"
```

The workaround which doesn't require root rights, is a symlink to your source directory:
```
ln -s "/Users/sarav/Sara Verbeecke Dropbox/My GitLab/micro_robot" "/Users/sarav/micro_robot"
```

In both workarounds use the directory `/Users/sarav/micro_robot` which does not contain spaces.

**Reply by Josverl** [reply]
Assuming your on windows wsl2 you may be able to use NTFS 8.3 naming
https://superuser.com/questions/129295/create-8-3-name-for-an-existing-directory
